File: Languages/cpp/algorithms-dsa/leetcode/sliding-window/minimum_window_substring.cpp
/*
 * LeetCode 76 - Minimum Window Substring
 * Topic: Sliding Window
 * Difficulty: Hard
 *
 * Expand window to include all chars of t, then contract to find minimum.
 * Use frequency maps and a "formed" counter.
 * Time: O(n)
 * Space: O(n)
 */
#include <iostream>
#include <string>
#include <unordered_map>
#include <climits>
#include <cassert>
using namespace std;

string minWindow(string s, string t) {
    unordered_map<char, int> need, have;
    for (char c : t) need[c]++;
    int required = need.size(), formed = 0;
    int left = 0, minLen = INT_MAX, minStart = 0;

    for (int right = 0; right < (int)s.size(); right++) {
        have[s[right]]++;
        if (need.count(s[right]) && have[s[right]] == need[s[right]])
            formed++;
        while (formed == required) {
            int len = right - left + 1;
            if (len < minLen) { minLen = len; minStart = left; }
            have[s[left]]--;
            if (need.count(s[left]) && have[s[left]] < need[s[left]])
                formed--;
            left++;
        }
    }
    return minLen == INT_MAX ? "" : s.substr(minStart, minLen);
}

int main() {
    assert(minWindow("ADOBECODEBANC", "ABC") == "BANC");
    assert(minWindow("a", "a") == "a");
    assert(minWindow("a", "aa") == "");
    assert(minWindow("aa", "aa") == "aa");

    cout << "All tests passed!" << endl;
    return 0;
}

File: Languages/python/algorithms-dsa/leetcode/sliding-window/minimum_window_substring.py
"""
LeetCode #76 - Minimum Window Substring
Topic: Sliding Window
Difficulty: Hard

Find minimum window in s containing all characters of t using expand/contract.

Time Complexity: O(n + m)
Space Complexity: O(m) where m = len(t)
"""
from collections import Counter


def min_window(s: str, t: str) -> str:
    if not t or not s:
        return ""
    need = Counter(t)
    have = 0
    required = len(need)
    left = 0
    result = ""
    min_len = float("inf")
    window: dict[str, int] = {}
    for right in range(len(s)):
        ch = s[right]
        window[ch] = window.get(ch, 0) + 1
        if ch in need and window[ch] == need[ch]:
            have += 1
        while have == required:
            if (right - left + 1) < min_len:
                min_len = right - left + 1
                result = s[left:right + 1]
            window[s[left]] -= 1
            if s[left] in need and window[s[left]] < need[s[left]]:
                have -= 1
            left += 1
    return result


if __name__ == "__main__":
    assert min_window("ADOBECODEBANC", "ABC") == "BANC"
    assert min_window("a", "a") == "a"
    assert min_window("a", "aa") == ""
    print("All tests passed!")
